WIP: Chore/docs blog (#925)
ing-bank/lion: docs/extending-documentation.md

> # Extending Lion Documentation

Reviewer: We have [testing static demo page](https://github.com/ing-bank/lion/pull/1010) hanging in the air, we should consider which if we first want to publish this, or wait till we are on the static demo page (and adapt this blog post to it).

(rewrite of original)
TODO: Check medium support for IDs in MarkDown
TODO: Add IDs
TODO: Make TOC
TODO: Consistent person

> Lion is a set of white label cross-platform Web Components you can use as the base for your own design system. You can extend them with your own styling for an accessible, performant and complete design system. 

Reviewer 2: I would also adjust this second sentence:
- it now looks a bit like we get a11y and perf as a a consequence of adding styles :) 
- reads a bit nicer if we don't end two consecutive sentences with `your own design system`

- [Read more about Lion](https://medium.com/ing-blog/ing-open-sources-lion-a-library-for-performant-accessible-flexible-web-components-22ad165b1d3d)
- [A demo of Lion](https://lion-web-components.netlify.app)
- [Lion on Github]()

> Extending Lion is not limited to extending components. Documentation and demos can be reused as well. This removes removes duplicate work such as writing and maintenance.

Reviewer 3: ```suggestion
Extending Lion is not limited to extending components. Documentation and demos can be reused as well. This removes duplicate work such as writing and maintaining documentation.
```

> This article explains how to extend both components and documentation.

Reviewer 3: ```suggestion
This article explains how to extend both components as well as documentation.
```

> > A naming convention that is similar to `Lion*` for class names and `lion-*` for tag names is required for this to work. We use the example of `Example` e.g. `ExampleButton` and `example-tabs`.

Reviewer 3: ```suggestion
> A naming convention that is similar to `Lion*` for class names and `lion-*` for tag names is required for this to work. For demoing purposes, we'll use the following names: `ExampleButton` and `example-tabs`.
```

## Table of Contents

> 0. Setting up, and Extending Lion

Reviewer 2: We don't refer to these numbers elsewhere. And should we start with 0?

1. Make a selection
2. Change input paths
3. Remove, add and replace sections
4. Conclusion

## Setting up, and extending Lion

This article assumes some basic terminal knowledge, and a working installation of npm. Yarn would work as well.

Create a new folder for our components using the terminal.

```sh
mkdir example-components
```

Enter the folder `example-components` and run the following command to scaffold a new project using [open-wc](https://open-wc.org).

```sh
npm init @open-wc
```

> When presented with a menu, (at least) pick the following options.

Reviewer 3: ```suggestion
When presented with a menu, pick (at least) the following options.
```

import { LionButton } from '@lion/button';

export class ExampleButton extends LionButton {
  static get styles() {
    return [
      super.styles,
      css`
        /* your styles can go here */
      `
    ];
  }

  connectedCallback() {
    super.connectedCallback();
    this._setupFeature();
  }

  _setupFeature() {
    // your code can go here
  }
}
```

You have now extended `<lion-button>` and created `<example-button>` from it. The component can be experienced in the browser by running `npm run storybook` inside the `example-button`-folder. 
Feel free to add styles and make it your own. This can be the start of a whole set of Web Components for your Design System.

For this article we assume you set up the project like mentioned before, using [Prebuilt Storybook](https://open-wc.org/demoing/) with [MDJS](https://open-wc.org/mdjs/). If you already have a repository, you can also add Storybook using [open-wc](https://open-wc.org). Enter the following:

```sh
npm init @open-wc
```

And pick 'upgrade an existing project'. Or install it manually by entering the following:

```sh
npm i @open-wc/demoing-storybook --save-dev
```

## Make a selection

> We need to specify which stories to load in `.storybook/main.js`.

Reviewer 3: I feel like this could use some more context, selection of what? why are we making a selection? what does storybook even do?

Change the following line:

```js
  stories: ['../stories/**/*.stories.{js,md,mdx}'],
```

to add the Lion readme

```js
  stories: ['../stories/**/*.stories.{js,md,mdx}', '../node_modules/@lion/button/README.md'],
```

> Here we extend the documentation of `LionButton`, for our own `ExampleButton`. This step alone should already give you the `LionButton` docs inside your own Storybook.

Reviewer 3: ```suggestion
This is where we extend the documentation of `LionButton`, for our own `ExampleButton`. This step alone should already give you the `LionButton` docs inside your own Storybook.
```

};
```

We import the providence output file (`providence-extend-docs-data.json'`) and pass it to the plugin options as the `changes` property.

The babel plugin runs for the files that we specify in the `test` property, replaces the imports, and replaces the tag names inside JavaScript code snippets!

> > It will only transforms JavaScript snippets that use [MDJS](https://open-wc.org/mdjs/) syntax such as \`\`\`js script, \`\`\`js story and \`\`\`js preview-story

Reviewer 3: ```suggestion
> It will only transform JavaScript snippets that use [MDJS](https://open-wc.org/mdjs/) syntax such as \`\`\`js script, \`\`\`js story and \`\`\`js preview-story
```

> We have to add Babel to our `es-dev-server`-configuration to make it run for Storybook.

Reviewer 3: ```suggestion
We also have to add Babel to our `es-dev-server`-configuration to make it work with Storybook.
```

};
```

We should now see the `LionButton` instances transformed into our own `ExampleButton`!

## Remove, add and replace sections

> You might not want to show all examples of how to use a component. Sometimes information is `Lion` specific, or perhaps in your design system people are not allowed to use a certain feature that we documented in `Lion`.

Reviewer 2: Minor detail: I think it's a common habit to write Design System in capitals (googling the term confirmed thi suspicion)

};
```

> In the code example mentioned, we have 2 places in where you can do overrides: `./.storybook/all.override.md` for generic overrides and via `getLocalOverridePath` for each component. When needed, the `rel` needs to be the same in `lion` and our own project to be able to override the right file.

Reviewer 2: Minor detail: numbers below 20 should ideally be written as text so (`two`) -> https://onzetaal.nl/taaladvies/getallen-in-letters-of-cijfers

In each file we need to specify which section we want to override. We want to load `example-button` in the project:

````md
```
::addMdAfter(':root')
```

```js script
import '../example-button.js';
```
````

And then replace each `button` with it.

````md
```js ::replaceFrom(':root')
module.exports.replaceSection = node => {
  if (node.type === 'code' && node.value) {
    let newCode = node.value;

>     newCode = newCode.replace(/<lion-button/g, '<example-button');

Reviewer 2: Tag replacement should have been taken care of by extend-docs and babel-extend, I think.

maybe we should also ask an external dev to 'do the tutorial'?

Author: That would be very nice. Got anybody in mind? Feel free to ask

    node.value = newCode;
  }
  return node;
};
```
````

> You can remove content by targetting a specific heading:

Reviewer 2: targetting -> targeting

````md
```
::removeFrom('heading:has([value=Usage with native form])')
```
````

Or you can add an extra paragraph below the content:

````md
```
::addMdAfter(':scope:last-child')
```

The documentation of [Remark extend](https://lion-web-components.netlify.app/?path=/docs/tools-remark-extend--page) has many more options and insights

> ## Conclusion

Reviewer 2: This text ends up inside the code block atm
